lib/realtime/repo_replica.ex

```diff
@@ -0,0 +1,75 @@
+defmodule Realtime.Repo.Replica do
+  @moduledoc """
+  Generates a read-only replica repo for the region specified in config/runtime.exs.
+  """
+  require Logger
+
+  @replicas_fly %{
+    "sea" => Realtime.Repo.Replica.SJC,
+    "sjc" => Realtime.Repo.Replica.SJC,
+    "gru" => Realtime.Repo.Replica.IAD,
+    "iad" => Realtime.Repo.Replica.IAD,
+    "sin" => Realtime.Repo.Replica.SIN,
+    "maa" => Realtime.Repo.Replica.SIN,
+    "syd" => Realtime.Repo.Replica.SIN,
+    "lhr" => Realtime.Repo.Replica.FRA,
+    "fra" => Realtime.Repo.Replica.FRA
+  }
+
+  @replicas_aws %{
+    "ap-southeast-1" => Realtime.Repo.Replica.Singapore,
+    "ap-southeast-2" => Realtime.Repo.Replica.Sydney,
+    "eu-west-1" => Realtime.Repo.Replica.Ireland,
+    "eu-west-2" => Realtime.Repo.Replica.London,
+    "us-east-1" => Realtime.Repo.Replica.NorthVirginia
+  }
+
+  @ast (quote do
+          use Ecto.Repo,
+            otp_app: :realtime,
+            adapter: Ecto.Adapters.Postgres,
+            read_only: true
+        end)
+  @doc """
+  Returns the replica repo module for the region specified in config/runtime.exs.
+  """
+  @spec replica() :: module()
+  def replica do
+    replicas =
+      case Application.get_env(:realtime, :platform) do
+        :aws -> @replicas_aws
+        :fly -> @replicas_fly
+        _ -> %{}
+      end
+
+    region = Application.get_env(:realtime, :region)
+    replica = Map.get(replicas, region)
+    replica_conf = Application.get_env(:realtime, replica)
+
+    # Do not create module if replica isn't set or configuration is not present
+    cond do
+      is_nil(replica) ->
+        Logger.warning("Replica region not found, defaulting to Realtime.Repo")
+        Realtime.Repo
+
+      is_nil(replica_conf) ->
+        Logger.error("Replica config not found for #{region} region")
+        Realtime.Repo
+
+      true ->
+        # Check if module is present
+        case Code.ensure_compiled(replica) do
+          {:module, _} -> nil
+          _ -> {:module, _, _, _} = Module.create(replica, @ast, Macro.Env.location(__ENV__))
+        end
```

Comment:
Is there a way to avoid dynamic creation of modules?

Reply:
trickier, we would need to define the region at compile time so we could properly load things up. with this change we're using runtime stuff which makes it trickier.

plus, with the current implementation, we are creating more repo modules as we are creating a Repo per region even if they are not used